algorithm.py

```diff
@@ -0,0 +1,140 @@
+
+        try:
+            neighbour = random.choices(neighbours, weights)
+        except IndexError:
+            neighbour = random.choices(neighbours)
+
+        return neighbour[0]
+
+    def move_ants(self, ant_matrix, pheromone_matrix, variance_matrix, alpha, beta):
+        for i, ant in enumerate(ant_matrix):
+            ant_matrix[i] = self.get_next_move(ant, pheromone_matrix, variance_matrix, alpha, beta)
+
+    @staticmethod
+    def init_pheromone_matrix(image_shape, initial_value):
+        """Return initial pheromone matrix filled with constant value.
+
+        """
+        pheromone_matrix = np.full((image_shape[0], image_shape[1]), initial_value)
+
+        return pheromone_matrix
+
+    @staticmethod
+    def pheromone_matrix_update(pheromone_matrix, ant_positions, heuristic_matrix, evaporation_rate):
+        """Return updated pheromone matrix after ants movement.
+
+        After every step, the pheromone values are updated. Evaporation rate controls the degree of the
+        updating of matrix.
+        """
+        for ant in ant_positions:
+            pheromone_matrix[ant[0], ant[1]] = (1 - evaporation_rate) * pheromone_matrix[ant[0], ant[1]] + \
+                                               evaporation_rate * heuristic_matrix[ant[0], ant[1]]
+
+    @staticmethod
+    def pheromone_matrix_decay(pheromone_matrix, initial_pheromone_matrix, pheromone_decay):
+        """ Return decayed pheromone matrix.
+
+        After the movement of all ants, the pheromone matrix is updated.
+        """
+        for i in range(0, pheromone_matrix.shape[0]):
+            for j in range(0, pheromone_matrix.shape[1]):
+                pheromone_matrix[i][j] = (1 - pheromone_decay) * pheromone_matrix[i][j] + pheromone_decay * \
+                                         initial_pheromone_matrix[i][j]
+
+    @staticmethod
+    def calculate_threshold(pheromone_matrix, epsilon):
+        """ Return a calculated threshold.
+
+        """
+        threshold = np.mean(pheromone_matrix)
+        i = 0
+        while True:
+            mi1 = np.mean(pheromone_matrix[pheromone_matrix >= threshold])
+            mi2 = np.mean(pheromone_matrix[pheromone_matrix < threshold])
+            new_threshold = (mi1 + mi2) / 2
+            if (abs(threshold - new_threshold) < epsilon) or (i > 1000):
+                break
+            i = i + 1
+        return new_threshold
+
+    def determine_edges(self, pheromone_matrix, epsilon):
+        im_edges = np.zeros(pheromone_matrix.shape, np.uint8)
+        thresh = self.calculate_threshold(pheromone_matrix, epsilon)
+        for i in range(0, pheromone_matrix.shape[0]):
+            for j in range(0, pheromone_matrix.shape[1]):
+                if pheromone_matrix[i][j] >= thresh:
+                    im_edges[i][j] = 255
+
+        return im_edges
+
+    def run_one_iteration(self, ant_matrix, pheromone_matrix):
+        self.move_ants(ant_matrix, pheromone_matrix, self.model.variance_matrix,
```

Comment:
Could you change ant_matrix & pheromone_matrix to be class' fields instead of function parameters?

Reply:
Done